### DISCUSSION TITLE
Function parsing with pipe

### DISCUSSION THREAD

**Discussion by broneill** [PRQL/prql#4290] [Q&A] [answered]
I'm experimenting with the parser, and something is confusing me.

`let a = 3 | math.pow 2`  # illegal -- unexpected | while parsing function call
`let a = (3 | math.pow 2)`  # legal

I don't understand why parens are needed in order for the expression to be parsed. In other contexts, the parens aren't needed.

**Comment by max-sixty** [comment, marked as answer]
Yes, the rules around parentheses are some of my least favorite rules. Check out https://prql-lang.org/book/reference/syntax/operators.html#parentheses...

The reason we can't always loosely bind `=` so that `let a = 3 | math.pow 2` parses to `let a =` & `3 | math.pow 2` is because that would parse `join foo=bar (==id)` to `join foo=` & `bar (==id)`, which isn't right. If we had different syntax for aliases vs. assignments, then we could bind `=` in the first example tighter, and require fewer parentheses. 

But this was the best tradeoff we could come up with. (Lots of old issues on this that are required context to make progress here...)

**Reply by broneill** [reply]
Perhaps the language grammar needs a complete review and possible overhaul? The version is at 0.11.4, still quite young.

**Reply by max-sixty** [reply]
This is a constraint of the current language. Check out https://github.com/PRQL/prql/issues/437 & links for past discussions. Starting to discuss a change would require examples, which take into account those past discussions...

**Reply by broneill** [reply]
I'm suggesting any specific changes, because I have no context. The review process is for the experts.
